File: src-tauri/src/types/enums.rs
use serde::{Deserialize, Serialize};
use specta::Type;

#[derive(Serialize, Deserialize, Type, Debug, PartialEq, Eq, Hash, Copy, Clone)]
#[serde(rename_all = "kebab-case")]
pub enum ScanResult {
    Clean,
    ThreatsFound,
    Partial,
    Failed,
    ClamavError
}

#[derive(Serialize, Deserialize, Type, Debug, PartialEq, Eq, Hash, Copy, Clone)]
#[serde(rename_all = "lowercase")]
pub enum ScanType {
    Main,
    Full,
    Custom,
    File,
    Realtime,
    Scheduled,
}

#[derive(Serialize, Deserialize, Type, Debug, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum ThreatStatus {
    Quarantined,
    Deleted,
    Unresolved,
}

#[derive(Serialize, Deserialize, Type, Debug, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum ComputerVirusType {
    Trojan,
    Ransomware,
    Spyware,
    Rootkit,
    Other,
}

#[derive(Serialize, Deserialize, Type, Debug, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum HistoryStatus {
    Success,
    Warning,
    Error,
    Acknowledged,
}

#[derive(Serialize, Deserialize, Type, Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum LogCategory {
    Scan,
    Update,
    Quarantine,
    Realtime,
    Scheduler,
}
impl LogCategory {
    pub fn as_str(&self) -> &'static str {
        match self {
            LogCategory::Scan => "scan",
            LogCategory::Update => "update",
            LogCategory::Quarantine => "quarantine",
            LogCategory::Realtime => "realtime",
            LogCategory::Scheduler => "scheduler",
        }
    }
}

#[derive(Serialize, Deserialize, Type, Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum SchedulerInterval {
    Daily,
    Weekly,
    Monthly,
}

#[derive(Serialize, Deserialize, Type, Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum DayOfTheWeek {
    Mon,
    Tue,
    Wed,
    Thu,
    Fri,
    Sat,
    Sun,
}
impl DayOfTheWeek {
    pub fn as_str(&self) -> &'static str {
        match self {
            DayOfTheWeek::Mon => "mon",
            DayOfTheWeek::Tue => "tue",
            DayOfTheWeek::Wed => "wed",
            DayOfTheWeek::Thu => "thu",
            DayOfTheWeek::Fri => "fri",
            DayOfTheWeek::Sat => "sat",
            DayOfTheWeek::Sun => "sun",
        }
    }
}

#[derive(Serialize, Deserialize, Type, Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[serde(rename_all = "lowercase")]
pub enum ClearHistoryMode {
    All,
    Acknowledged,
    Error,
    Warning
}

#[derive(Serialize, Deserialize, Type, Debug, PartialEq, Eq, Hash, Copy, Clone)]
#[serde(rename_all = "lowercase")]
pub enum BehaviorMode {
    Balanced,
    Safe,
    Strict,
    Expert,
}

#[derive(Serialize, Deserialize, Type, Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[serde(rename_all = "kebab-case")]
pub enum HistoryType{
    RealTimeError,
    RealTimeStart,
    RealTimeStop,
    QuarantineThreat,
    RestoreThreat,
    DeleteThreat,
    ScanStart,
    ScanFinish,
    DefUpdateStart,
    DefUpdateFinish,
    DefUpdateError,
    SchedulerCreate,
    SchedulerDelete,
    SchedulerTrigger,
    FileDelete,
    FileDeleteError
}

#[derive(Debug, Serialize, Deserialize, Type)]
#[serde(tag = "type", content = "details", rename_all = "kebab-case")]
pub enum HistoryDetails{
    RealTimeError {err: String},
    RealTimeStart {behavior: BehaviorMode, paths: usize},
    RealTimeStop,
    QuarantineThreat {threat: String},
    RestoreThreat {threat: String},
    DeleteThreat {threat: String},
    ScanStart {scan_type: ScanType},
    ScanFinish {result: ScanResult, exit_code: i32, found_threats: usize},
    DefUpdateStart,
    DefUpdateFinish {exit_code: i32},
    DefUpdateError {err: String},
    SchedulerCreate {task_name: String},
    SchedulerDelete {task_name: String},
    SchedulerTrigger {task_name: String},
    SchedulerTriggerError {task_name: String},
    FileDelete {file_path: String},
    FileDeleteError {err: String, file_path: String}
}
#[derive(Serialize, Deserialize, Type, Debug, PartialEq, Eq, Hash, Copy, Clone)]
#[serde(rename_all = "lowercase")]
pub enum RealTimeState{
    Enabled,
    Disabled,
    Default
}

#[derive(Serialize, Deserialize, Type, Debug, PartialEq, Eq, Hash, Copy, Clone)]
#[serde(rename_all = "camelCase")]
pub enum SettingKeyArray{
    Exclusions,
    MonitoringPaths
}
impl SettingKeyArray {
    pub fn as_str(&self) -> &'static str {
        match self {
            SettingKeyArray::Exclusions => "exclusions",
            SettingKeyArray::MonitoringPaths => "monitoringPaths"
        }
    }
}

File: src/lib/types/enums.ts
export enum QuickAccessLink {
     MainScan = "main-scan",
     FullScan = "full-scan",
     CustomScan = "custom-scan",
     FileScan = "file-scan",
     Quarantine = "quarantine",
     Update = "update",
     History = "history",
     RealTime = "real-time"
}
export enum SidebarLink {
     Overview = "overview",
     Scan = "scan",
     Quarantine = "quarantine",
     History = "history",
     Stats = "stats",
     Scheduler = "scheduler",
     Settings = "settings",
     About = "about"
}
export enum Indicator {
     Safe = "safe",
     Warning = "warning",
     Alert = "alert",
}
export enum DefinitionStatus {
     Updated = "updated",
     Outdated = "outdated",
     Loading = "loading"
}
export enum LogCategory {
     Scan = "scan",
     Update = "update",
     Quarantine = "quarantine",
     RealTime = "realtime",
     Scheduler = "scheduler"
}
export enum SettingsTab {
     General = "general",
     Advanced = "advanced",
     Scan = "scan",
     Update = "update"
}
export enum DateFormatType {
     American = "us",
     European = "eu",
     International = "iso"
}
export enum ClamAVState {
     Checking = "checking",
     Available = "available",
     Missing = "missing"
}
export enum GuiUpdaterStatus {
     Checking = "checking",
     Updating = "updating",
     NeedsUpdate = "needs-update",
     Updated = "updated",
     CheckError = "failed-check",
     Completed = "completed",
     UpdateError = "failed-update"
}
export enum ScanType {
     None = "",
     Main = "main",
     Full = "full",
     Custom = "custom",
     File = "file"
}
export enum ScanProfile {
     Main = "main",
     Custom = "custom",
     File = "file"
}
export enum BehaviorMode {
     Balanced = "balanced",
     Safe = "safe",
     Strict = "strict",
     Expert = "expert"
}
export enum BadgeVisibility {
     Icon = "icon",
     IconText = "icon-text",
     Text = "text"
}
export enum ScanOptionGroup {
     Detection = "detection",
     FileTypes = "file-types",
     FileSystem = "filesystem",
     LimitsPerformance = "limits-performance",
     Output = "output",
     Advanced = "advanced"
}
export enum HistoryClearType {
     All = "all",
     Acknowledged = "acknowledged",
     Error = "error",
     Warning = "warning"
}
export enum HistoryType {
     RealTimeError = "real-time-error",
     RealTimeStart = "real-time-start",
     RealTimeStop = "real-time-stop",
     QuarantineThreat = "quarantine-threat",
     RestoreThreat = "restore-threat",
     DeleteThreat = "delete-threat",
     ScanStart = "scan-start",
     ScanFinish = "scan-finish",
     DefUpdateStart = "def-update-start",
     DefUpdateFinish = "def-update-finish",
     DefUpdateError = "def-update-error",
     SchedulerCreate = "scheduler-create",
     SchedulerDelete = "scheduler-delete",
     SchedulerTrigger = "scheduler-trigger",
     FileDelete = "file-delete",
     FileDeleteError = "file-delete-error"
}
export enum ScanResult {
    Clean = "clean",
    ThreatsFound = "threats-found",
    Partial = "partial",
    Failed = "failed",
    ClamavError = "clamav-error"
}